File: Android/Beacon/app/src/main/java/com/fissoft/app/MainActivity.kt
package com.fissoft.app

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.IBinder
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat

class MainActivity : AppCompatActivity() {
    private val connection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
        }

        override fun onServiceDisconnected(name: ComponentName?) {
        }
    }

    private val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ),
                233
            )
        } else {
            bindService(
                Intent(this, MyService::class.java),
                connection,
                Context.BIND_AUTO_CREATE
            )
        }

    }


    override fun onDestroy() {
        super.onDestroy()
        unbindService(connection)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 233 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            bindService(
                Intent(this, MyService::class.java),
                connection,
                Context.BIND_AUTO_CREATE
            )
        } else {
            Toast.makeText(this, "用户拒绝了权限,无法定位", Toast.LENGTH_LONG).show()
        }
    }
}

File: App/MainActivity.cs
using Android;
using Android.App;
using Android.Bluetooth;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Widget;
using AndroidX.AppCompat.App;
using AndroidX.Core.App;
using System;

namespace App
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme", MainLauncher = true)]
    public class MainActivity : AppCompatActivity, BluetoothAdapter.ILeScanCallback
    {

        private BluetoothAdapter bluetoothAdapter = BluetoothAdapter.DefaultAdapter;
        private readonly BeaconServiceConnection beaconServiceConnection = new BeaconServiceConnection();
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Essentials.Platform.Init(this, savedInstanceState);
            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.activity_main);

            var btnScan = this.FindViewById<Button>(Resource.Id.btn_scan);
            btnScan.Click += (s, e) =>
            {
                bluetoothAdapter.StartLeScan(this);
            };

            if (ActivityCompat.CheckSelfPermission(this, Manifest.Permission.AccessFineLocation)
                != Android.Content.PM.Permission.Granted
            )
            {
                ActivityCompat.RequestPermissions(
                    this,
                   new string[] { Manifest.Permission.AccessFineLocation, Manifest.Permission.AccessCoarseLocation },
                    233
                );
            }
            else
            {
                BindService(
                   new Intent(this, typeof(BeaconService)),
                beaconServiceConnection,
               Bind.AutoCreate
            );
            }
        }

        public override void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Android.Content.PM.Permission[] grantResults)
        {
            base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
            if (requestCode == 233 && grantResults[0] == Android.Content.PM.Permission.Granted)
            {
                BindService(
                   new Intent(this, typeof(BeaconService)),
                beaconServiceConnection,
               Bind.AutoCreate
            );
            }
            else
            {
                Toast.MakeText(this, "用户拒绝了权限,无法定位", ToastLength.Long).Show();
            }
        }

        public void OnLeScan(BluetoothDevice device, int rssi, byte[] scanRecord)
        {
            double d = Math.Pow(10.0, (Math.Abs(rssi) - 59) / (10 * 2.0));
            Android.Util.Log.Info("OnLeScan", $"onCreate: {device.Address} 信号强度：{rssi}  距离：{d}");
            if (device.Address == "D8:96:E0:8B:8B:72")
            {
                bluetoothAdapter.StopLeScan(this);
            }
        }
    }
}